pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/DateTruncTransformFunctionTest.java

```diff
@@ -0,0 +1,172 @@
+  private static final String TIMESTAMP_ISO8601_STRING = "2001-08-22T03:04:05.321+00:00";
+
+  private static final DateTime TIMESTAMP = new DateTime(2001, 8, 22, 3, 4, 5, 321, UTC_TIME_ZONE); // This is TIMESTAMP w/o TZ
+
+
+  private static long iso8601ToUtcEpochMillis(String iso8601) {
+    DateTimeFormatter formatter = ISODateTimeFormat.dateTimeParser().withOffsetParsed();
+    return formatter.parseDateTime(iso8601).getMillis();
+  }
+
+  private static void testDateTruncHelper(String literalInput, String unit, String tz, long expected) throws Exception {
+    long zmillisInput = iso8601ToUtcEpochMillis(literalInput);
+    GenericRow row = new GenericRow();
+    row.init(ImmutableMap.of(TIME_COLUMN, zmillisInput));
+    List<GenericRow> rows = ImmutableList.of(row);
+    Schema schema = new Schema();
+    schema.addField(new TimeFieldSpec(TIME_COLUMN, FieldSpec.DataType.LONG, TimeUnit.MILLISECONDS));
+
+    SegmentGeneratorConfig config = new SegmentGeneratorConfig(schema);
+    String segmentName = "testSegment";
+    String indexDirPath = Paths.get(Files.createTempDirectory("pinot_date_trunc_test").toAbsolutePath().toString(), segmentName).toAbsolutePath().toString();
+    try {
+      FileUtils.deleteQuietly(new File(indexDirPath));
+      config.setOutDir(indexDirPath);
+      config.setSegmentName(segmentName);
+      SegmentIndexCreationDriverImpl driver = new SegmentIndexCreationDriverImpl();
+      driver.init(config, new GenericRowRecordReader(rows, schema));
+      driver.build();
+
+      IndexSegment indexSegment = ImmutableSegmentLoader.load(new File(indexDirPath, segmentName), ReadMode.heap);
+      Set<String> columnNames = indexSegment.getPhysicalColumnNames();
+      HashMap<String, DataSource> dataSourceMap = new HashMap<>(columnNames.size());
+      for (String columnName : columnNames) {
+        dataSourceMap.put(columnName, indexSegment.getDataSource(columnName));
+      }
+
+      ProjectionBlock projectionBlock = new ProjectionOperator(dataSourceMap,
+              new DocIdSetOperator(new MatchAllFilterOperator(rows.size()), DocIdSetPlanNode.MAX_DOC_PER_CALL)).nextBlock();
+
+      TransformExpressionTree expression = TransformExpressionTree.compileToExpressionTree(
+              String.format("dateTrunc(%s, '%s', '%s', '%s')", TIME_COLUMN, TimeUnit.MILLISECONDS, tz, unit));
+      TransformFunction transformFunction = TransformFunctionFactory.get(expression, dataSourceMap);
+      Assert.assertTrue(transformFunction instanceof DateTruncTransformFunction);
+      Assert.assertEquals(transformFunction.getName(), DateTruncTransformFunction.FUNCTION_NAME);
+      long[] longValues = transformFunction.transformToLongValuesSV(projectionBlock);
+      Assert.assertEquals(longValues[0], expected);
+    } finally {
+      FileUtils.deleteDirectory(new File(indexDirPath));
+    }
+  }
+
+  @Test
+  public void testPrestoCompatibleDateTimeConversionTransformFunction() throws Exception {
```

Comment:
Since dateTrunc() function can accept transform function as an argument, tt would be good to add tests that verify the nested nature. You can look at TransformQueriesTest for example.

Reply:
Oh cool. So I will add the test for this new UDF also there.